logpy/tests/test_sudoku.py

```diff
@@ -0,0 +1,165 @@
+    )
+
+
+def test_missing_one_entry():
+    example_board = (
+        5, 3, 4, 6, 7, 8, 9, 1, 2,
+        6, 7, 2, 1, 9, 5, 3, 4, 8,
+        1, 9, 8, 3, 4, 2, 5, 6, 7,
+        8, 5, 9, 7, 6, 1, 4, 2, 3,
+        4, 2, 6, 8, 5, 3, 7, 9, 1,
+        7, 1, 3, 9, 2, 4, 8, 5, 6,
+        9, 6, 1, 5, 3, 7, 2, 8, 4,
+        2, 8, 7, 4, 1, 9, 6, 3, 5,
+        3, 4, 5, 2, 8, 6, 0, 7, 9
+    )
+    expected_solution = (
+        5, 3, 4, 6, 7, 8, 9, 1, 2,
+        6, 7, 2, 1, 9, 5, 3, 4, 8,
+        1, 9, 8, 3, 4, 2, 5, 6, 7,
+        8, 5, 9, 7, 6, 1, 4, 2, 3,
+        4, 2, 6, 8, 5, 3, 7, 9, 1,
+        7, 1, 3, 9, 2, 4, 8, 5, 6,
+        9, 6, 1, 5, 3, 7, 2, 8, 4,
+        2, 8, 7, 4, 1, 9, 6, 3, 5,
+        3, 4, 5, 2, 8, 6, 1, 7, 9,
+    )
+    assert sudoku_solver(example_board)[0] == expected_solution
+
+
+def test_missing_complex_board():
+    example_board = (
+        5, 3, 4, 6, 7, 8, 9, 0, 2,
+        6, 7, 2, 0, 9, 5, 3, 4, 8,
+        0, 9, 8, 3, 4, 2, 5, 6, 7,
+        8, 5, 9, 7, 6, 0, 4, 2, 3,
+        4, 2, 6, 8, 5, 3, 7, 9, 0,
+        7, 0, 3, 9, 2, 4, 8, 5, 6,
+        9, 6, 0, 5, 3, 7, 2, 8, 4,
+        2, 8, 7, 4, 0, 9, 6, 3, 5,
+        3, 4, 5, 2, 8, 6, 0, 7, 9,
+    )
+    expected_solution = (
+        5, 3, 4, 6, 7, 8, 9, 1, 2,
+        6, 7, 2, 1, 9, 5, 3, 4, 8,
+        1, 9, 8, 3, 4, 2, 5, 6, 7,
+        8, 5, 9, 7, 6, 1, 4, 2, 3,
+        4, 2, 6, 8, 5, 3, 7, 9, 1,
+        7, 1, 3, 9, 2, 4, 8, 5, 6,
+        9, 6, 1, 5, 3, 7, 2, 8, 4,
+        2, 8, 7, 4, 1, 9, 6, 3, 5,
+        3, 4, 5, 2, 8, 6, 1, 7, 9,
+    )
+    assert sudoku_solver(example_board)[0] == expected_solution
+
+
+def test_insolvable():
+    example_board = (
+        5, 3, 4, 6, 7, 8, 9, 1, 2,
+        6, 7, 2, 1, 9, 5, 9, 4, 8,  # Note column 7 has two 9's.
+        1, 9, 8, 3, 4, 2, 5, 6, 7,
+        8, 5, 9, 7, 6, 1, 4, 2, 3,
+        4, 2, 6, 8, 5, 3, 7, 9, 1,
+        7, 1, 3, 9, 2, 4, 8, 5, 6,
+        9, 6, 1, 5, 3, 7, 2, 8, 4,
+        2, 8, 7, 4, 1, 9, 6, 3, 5,
+        3, 4, 5, 2, 8, 6, 0, 7, 9
+    )
+    assert sudoku_solver(example_board) == ()
+
+
+def test_many_missing_elements():
+    example_board = (
+        5, 3, 0, 0, 7, 0, 0, 0, 0,
+        6, 0, 0, 1, 9, 5, 0, 0, 0,
+        0, 9, 8, 0, 0, 0, 0, 6, 0,
+        8, 0, 0, 0, 6, 0, 0, 0, 3,
+        4, 0, 0, 8, 0, 3, 0, 0, 1,
+        7, 0, 0, 0, 2, 0, 0, 0, 6,
+        0, 6, 0, 0, 0, 0, 2, 8, 0,
+        0, 0, 0, 4, 1, 9, 0, 0, 5,
+        0, 0, 0, 0, 8, 0, 0, 7, 9)
+    assert sudoku_solver(example_board)[0] == (
+        9, 3, 8, 2, 5, 6, 4, 1, 7,
+        2, 1, 4, 3, 7, 9, 8, 6, 5,
+        5, 7, 6, 8, 1, 4, 3, 2, 9,
+        4, 5, 2, 1, 3, 7, 9, 8, 6,
+        1, 6, 7, 5, 9, 8, 2, 4, 3,
+        8, 9, 3, 4, 6, 2, 5, 7, 1,
+        3, 2, 1, 7, 4, 5, 6, 9, 8,
+        7, 8, 9, 6, 2, 3, 1, 5, 4,
+        6, 4, 5, 9, 8, 1, 7, 3, 2,
+    )
+
+
+@pytest.mark.skip(reason="Currently too slow!")
+def test_websudoku_easy():
+    # A sudoku from websudoku.com.
+    example_board = (
```

Comment:
By comparison this solver based on `python-constraint` solves this sudoku in a small fraction of a second: https://gist.github.com/lucaswiman/f6769d2e866407dd784d1f29d3556771

The issue seems to be that due to goal reordering in `lall`, it ends up just computing a naive cartesian product, then checking if it solves the Sudoku. It would probably be difficult to solve this in a reasonable amount of time without implementing something like `clpfd` in Prolog or `fd` in `core.logic`.